sensorserviceutils/sensorserviceutils.go

```diff
@@ -0,0 +1,402 @@
+}
+
+// SubscriptionInfo contains information about one subscription event of
+// a sensor to an application.
+// For NRD42 and onwards Android versions: each subscription event is captured
+// by the + statement that adds the subscription and the - statement that
+// removes the subscription .
+// For MNC or before: each subscription event is captured by
+// activated/de-activated statments.
+type SubscriptionInfo struct {
+	StartMs, EndMs int64
+	SensorNumber   int32
+	UID, PID       int32
+	PackageName    string
+	SamplingPeriod int32
+	BatchingPeroid int32
+}
+
+type parser struct {
+	// referenceYear is the year extracted from the dumpstate line in a bugreport.
+	// Previous Registration lines don't contain a year in the date string,
+	// so we use this to reconstruct the full timestamp.
+	referenceYear int
+
+	// referenceMonth is the month extracted from the dumpstate line in a bugreport.
+	// Since a bugreport may span over a year boundary, we use the month to
+	// check whether the year for the event needs to be decremented or incremented.
+	referenceMonth time.Month
+
+	// referenceDay is the month extracted from the dumpstate line in a bugreport.
+	referenceDay int
+
+	// loc is the location parsed from timezone information in the bugreport.
+	// The previous registration is in the user's local timezone
+	// which we need to convert to UTC time.
+	loc *time.Location
+
+	lines []string
+	idx   int
+
+	buf         *bytes.Buffer
+	csvState    *csv.State
+	parsingErrs []error
+	sensorErrs  []error
+
+	// sensors is a map from sensor number to the relevant sensor's information.
+	sensors map[int32]bugreportutils.SensorInfo
+
+	// activeConns is a map from an identifier to the relevant connection information.
+	// If a sensor is actively subscribed by a package when the bugreport is
+	// generated, the relevant connection information can be obtained using
+	// an identifier formed by concatenating sensor number and package name.
+	activeConns map[string]*acpb.ActiveConn
+
+	// history is a map from an identifier to an sensor subscription event.
+	// Note that the identifier is a string formed by concatenating
+	// sensor number and name of the package that subscribes the sensor.
+	history map[string]*SubscriptionInfo
+}
+
+// Returns the current line without advancing the line position.
+func (p *parser) peek() string {
+	if p.valid() {
+		return p.lines[p.idx]
+	}
+	return ""
+}
+
+// Returns the current line and advances the line position.
+func (p *parser) line() string {
+	if !p.valid() {
+		return ""
+	}
+	l := p.lines[p.idx]
+	p.idx++
+	return l
+}
+
+// Returns the previous line and move to that line position.
+func (p *parser) prevline() string {
+	if !p.valid() {
+		return ""
+	}
+	if p.idx <= 0 {
+		return ""
+	}
+	p.idx--
+	l := p.lines[p.idx]
+	return l
+}
+
+// Returns whether the current line position corresponds to a valid line.
+func (p *parser) valid() bool {
+	return (p.idx < len(p.lines)) && (p.idx >= 0)
```

Comment:
Are there any other implementations for parsing in the tool that already exist?

Reply:
There are mainly two ways of parsing in the tool. 

One implementation can be found in the activity.go file. Since all the input that is given to the activity parser follows the logcat format, this parsing implementation calls the same helper function to parse one line every time.

The one I am using now follows the implementation for the parser in the broadcast.go file. I am using the broadcast parser since the sensorservice dump has a specific structure where the "sensor list" section comes first, followed by the "sensor events" section, "active connection" section, etc. 
In this case, I can define helper functions that handle each section: As the parser reads the sensor service dump, when we reach one section, we start parsing all information in this section using the corresponding function until we move on to the next section. 
I am using this parsing implementation since it shows better modularity in the design.